### PR TITLE
feat_delete

### PR DESCRIPTION
**Pull request by Gresham429**
Finish delete in HNSW, add LSH search with a successful test.
TODO: 1. Trying to migrate NSG  in this demo. 2. LSHAPG seemed to be crushed with its delete function. I suppose that might be the fault of its hash table's maintenance

### CODE REVIEW

**Comment by ShuhaoZhangTony** [comment]
Hi, would you like to join our slack channel for easier communication?

File: bin/Audio.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
Comment (ShuhaoZhangTony) [inline bin/Audio.sh:1]: I think we shall integrate the scripts for downloading data.

File: include/CANDY/FlatIndex.h [hunk 2]
```diff
@@ -97,6 +97,12 @@ class FlatIndex : public AbstractIndex {
    * @return a vector of tensors, each tensor represent KNN results of one query in idx
    */
   virtual std::vector<torch::Tensor> getTensorByIndex(std::vector<faiss::idx_t> &idx, int64_t k);
+  /**
+   * @brief delete a tensor by index
+   * @param idx the index
+   * @return a vector of tensors, each tensor represent KNN results of one query in idx
+   */
+  virtual bool deleteTensorByIndex(int64_t idx);
```
Comment (ShuhaoZhangTony) [inline include/CANDY/FlatIndex.h:105]: I can't see a use case that we shall delete vertext by id -- how is the id to be deleted known by user?

File: include/CANDY/LSH.h
Comment (ShuhaoZhangTony) [inline include/CANDY/LSH.h:1]: Misplaced?

File: include/CANDY/FlatIndex.h (hunk 2 above)
Comment (Rubatozh) [reply include/CANDY/FlatIndex.h:105]: Suppose the user would like to delete specific contexts. In that case, he should be able to search for IDs whose vectors are similar to the query to be deleted and then perform deleteTensorByIndex() using these IDs.
BTW in FreshDiskANN and SPFresh, their experiments delete a random portion of the dataset.

Comment (Gresham429) [reply include/CANDY/FlatIndex.h:105]: In fact, LSHAPG use FlatIndex to store the Tensor,  when we trying to delete the tensor, we just search its index, and use index to delete it from memory.
BTW, the hash table maintenance in the LSHAPG algorithm is incorrect.

File: bin/Audio.sh (hunk 1 above)
Comment (Gresham429) [reply bin/Audio.sh:1]: Okay, I'll do this after I complete my experiment.

File: include/CANDY/LSH.h
Comment (Gresham429) [reply include/CANDY/LSH.h:1]: ? I suppose it's right.

File: include/CANDY/FlatIndex.h (hunk 2 above)
Comment (Rubatozh) [reply include/CANDY/FlatIndex.h:105]: > In fact, LSHAPG use FlatIndex to store the Tensor, when we trying to delete the tensor, we just search its index, and use index to delete it from memory. BTW, the hash table maintenance in the LSHAPG algorithm is incorrect.

Would you provide us with more details about the incorrect hash table maintenance?

File: include/CANDY/LSH.h
Comment (Rubatozh) [reply include/CANDY/LSH.h:1]: > ? I suppose it's right.

Renaming LSH.cpp LSH.h and directory LSH to IndexLSH* should be fine

File: include/CANDY/FlatIndex.h (hunk 2 above)
Comment (Gresham429) [reply include/CANDY/FlatIndex.h:105]: In LSHAPG,  we load our initial tensor firstly,  we maintenace an hashTables to record the key and id, but when going to append a tensor, I can't see where is the table maintenance. 

image
Function appendHash do not operate the hash table.

Comment (Rubatozh) [reply include/CANDY/FlatIndex.h:105]: Screen Shot 2024-10-09 at 18 32 36
HashTables are updated here in insertLSHRefine(), which is called in oneByOneInsert() and appendTensor() when building and updating the index.

Comment (Gresham429) [reply include/CANDY/FlatIndex.h:105]: thanks!!! BTW, function appendHash did what in appendTensor

Comment (Rubatozh) [reply include/CANDY/FlatIndex.h:105]: Please refer to section 4.2 of [LSHAPG paper](https://www.vldb.org/pvldb/vol16/p1979-zhao.pdf). Basically there are L B+Trees, each with K hash functions. So there are L*K hashes in total, and appendHash computes L*K*newDataSize hash values for further use in the graph.

Comment (Gresham429) [reply include/CANDY/FlatIndex.h:105]: Thanks!!!!! I got it

**Pull request closed without merging**